Split Actor.Bounds into RenderBounds and SelectableBounds

reaperrr opened the pull request:
Additionally, internally renamed `VisualBounds` to `SelectionOverlayBounds` to avoid confusion with `RenderBounds`.

This step was necessary to prevent actors with selectable area smaller than their graphics to be removed too early from ScreenMap even though part of the graphics should still be visible.
RA cruisers were a prime example, but to a lesser extent several other actors were affected as well.

This also serves as preparation to determine the final `RenderBounds` from multiple source bounds without worrying about the selectable area, to fix the remaining ScreenMap issues (building 'bibs', aircraft shadows, parachuted infantry shadows and possibly a few more edge cases).

reaperrr committed:
Split Actor.Bounds into RenderBounds and SelectableBounds

Additionally, internally renamed VisualBounds to SelectionOverlayBounds to avoid confusion with RenderBounds.

This step was necessary to prevent actors with selectable area smaller than their graphics to be removed too early from ScreenMap even though part of the graphics should still be visible.
RA cruisers were a prime example, but to a lesser extent several other actors were affected as well.

This separation also serves as preparation to determine the final RenderBounds from multiple source bounds later, to fix the remaining ScreenMap issues (building 'bibs', aircraft shadows).

pchote commented on OpenRA.Game/Actor.cs:
```diff
@@ -47,8 +47,9 @@ internal struct SyncHash
-		public Rectangle VisualBounds { get; private set; }
+		public Rectangle RenderBounds { get; private set; }
+		public Rectangle SelectableBounds { get; private set; }
+		public Rectangle SelectionOverlayBounds { get; private set; }
```
Can you please move this to the `Selectable` trait?  The other two are needed by engine level code, but this one is purely visual.

One of the things on my Next+1 todo list is to rework the Selectable trait to support mod-defined selection boxes and non-rectangular click regions.  The changes here otherwise nicely prepare for this.

reaperrr replied:
> Can you please move this to the Selectable trait? The other two are needed by engine level code, but this one is purely visual.

This requires quite a bunch of changes and `Selectable` trait look-ups, of which at least the one in `SelectionBarsRenderable` can't be cached, as far as I can tell. Are you sure?

pchote replied:
You should be able to pass the trait to the constructor alongside the actor.

reaperrr replied:
Leaving as-is for now as discussed in IRC.

pchote commented on OpenRA.Mods.Common/Traits/Render/RenderNameTag.cs:
```diff
@@ -47,7 +47,7 @@ public RenderNameTag(Actor self, RenderNameTagInfo info)
 		{
 			var pos = wr.ScreenPxPosition(self.CenterPosition);
-			var bounds = self.Bounds;
+			var bounds = self.SelectableBounds;
```
Can you please change this to `SelectionOverlayBounds`?

reaperrr committed:
Move SelectionBarsRenderable to Mods.Common

reaperrr committed:
Switch RenderNameTag to use SelectionOverlayBounds

reaperrr commented:
Updated.

Fixed a typo, migrated `RenderNameTag` to `SelectionOverlayBounds` and moved `SelectionBarsRenderable` to Mods.Common (whatever dependency prevented that previously seems to be gone).

pchote commented on OpenRA.Game/Actor.cs:
```diff
@@ -129,24 +136,34 @@ internal Actor(World world, string name, TypeDictionary initDict)
-		Rectangle DetermineBounds()
+		Rectangle DetermineRenderBounds()
+		{
+			var size = TraitsImplementing<IAutoSelectionSize>().Select(x => x.SelectionSize(this)).FirstOrDefault();
```
Using `IAutoSelectionSize` for the render bounds feels really wrong.  Can we introduce a new interface, even if it just returns the same values for now?

reaperrr committed:
Use new IAutoRenderSize for determining actor render bounds

reaperrr commented:
Updated.

pchote commented:
This appears to have broken tunnel tops in TS.

reaperrr commented:
> This appears to have broken tunnel tops in TS.

Updated, I had forgotten about `CustomSelectionSize`.

pchote commented:
Hrm, that's not very nice.  Can we make `WithSpriteBody` implement `IAutoRenderSize` instead, and then drop that from `AutoSelectionSize` and `CustomSelectionSize`?

pchote commented:
Or at least, for now, have a separate `CustomRenderSize` trait if you want to minimize scope-creeping into what you wanted to do next wrt fixing bibs.

reaperrr committed:
Remove superflous AutoSelectionSize from TS high bridge placeholder

It defines CustomSelectionSize, so AutoSelectionSize is redundant/wrong here.

reaperrr commented:
The first option was less work anyway, so I went with that.

pchote commented:
It unfortunately looks like the first option wasn't good enough... There are a few actors that use `CustomSelectionSize` (e.g. TS tunnel tops) which now fall back to their incorrect automatic render size and glitch out when near the edge of the screen :sad:

reaperrr commented:
Updated.

penev92 commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -325,6 +326,7 @@
     <Compile Include="Traits\Crushable.cs" />
     <Compile Include="Traits\CustomSellValue.cs" />
     <Compile Include="Traits\CustomSelectionSize.cs" />
+    <Compile Include="Traits\CustomRenderSize.cs" />
```
Is there a reason this isn't in `Traits\Render\` with `AutoRenderSize`?

reaperrr replied:
Mostly that I wanted to have the Render counterpart in the same folder as the Selection trait, and `CustomSelectionSize` is now really unrelated to rendering.

penev92 commented on OpenRA.Mods.Common/Traits/Render/WithVoxelBody.cs:
```diff
@@ -39,7 +39,7 @@ public class WithVoxelBodyInfo : ConditionalTraitInfo, IRenderActorPreviewVoxels
 	}
 
-	public class WithVoxelBody : ConditionalTrait<WithVoxelBodyInfo>, IAutoSelectionSize
+	public class WithVoxelBody : ConditionalTrait<WithVoxelBodyInfo>, IAutoSelectionSize, IAutoRenderSize
```
You create `AutoRenderSize` as a separate trait for sprite actors, but implement the interface directly here.
For consistency we should go with either a second trait that works with `WithVoxelBody` the way `AutoRenderSize` works with `RenderSprites`, or get rid of `AutoRenderSprites` and implement the interface directly.

reaperrr replied:
Implementing the interface directly on WithSpriteBody or RenderSprites would be too problematic (since there must be only 1 trait with that interface per actor) and was dropped for that reason (this was also the main motivation when making AutoSelectionSize and CustomSelectionSize separate traits).

I see your point about consistency, but if possible I don't want to do the split for voxels in this PR yet, since that would make the PR and upgrade rules grow further without any practical benefit. We only really need the `Auto-/CustomRenderSize` traits for sprites, since there are several special cases that would otherwise be hard to get right, while I don't know of any voxels that would need this.

This is all about laying the groundwork to fix issues like bibs plopping in and out of the screen because they don't count towards the render bounds.

penev92 commented:
Also needs a rebase, but otherwise looks good.

reaperrr commented:
Rebased and updated.

penev92 reviewed (approved):
Not going to hold this over that inconsistency, because we already have the same thing for a number of other interfaces.
So 👍

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ee0c5737d402d6530e44f3a32f6393e183900ba6...174a0a37a30cee58f1f31636e626f393f26d8ed7
